light/odr.go

```diff
@@ -119,6 +119,23 @@ func (req *BlockRequest) StoreResult(db ethdb.Database) {
 	rawdb.WriteBodyRLP(db, req.Hash, req.Number, req.Rlp)
 }
 
+type hashOrNumber struct {
+	Hash   common.Hash
+	Number uint64
+}
+
+type HeaderRequest struct {
+	OdrRequest
+	Origin hashOrNumber
+	Header *types.Header
+}
+
+func (req *HeaderRequest) StoreResult(db ethdb.Database) {
```

Comment:
Seems like the function `HeaderChain.WriteHeader` does a lot of the same functionality as this function.  How are they different?

Reply:
Don't have access to `HeaderChain` here

Reply:
Are the `StoreResult` writes also done here:  https://github.com/celo-org/celo-blockchain/pull/1031/files#diff-3e970814285a79d487cdf2e12822e0bcR244 in `client_handler`?

If so, would it make sense to remove the writes in `StoreResult`?

Reply:
Yes this looks right, good catch. I'll modify `storeResult`

Reply:
I'm actually now wondering if the write to the header chain and write to the canonical hash store needs to be atomic.  I'd presume not (since I doubt it's done in upstream geth), but I think it would be worth looking into it a bit of what happens if they are inconsistent.  E.g. what happens if one is written and the other isn't (say from a client crash).

Reply:
Yeah this is a good point, this does introduce the potential for inconsistent reads - the impact seems small but not 0.
Wonder if I should look into updating `InsertHeaderChain` or if there's another solution, will keep investigating

Reply:
Seems like there is an expectation that the header is in the chaindb if the canonicalhash is in the db.

https://github.com/celo-org/celo-blockchain/blob/master/light/odr_util.go#L35

Does your changes maintain that constraint?

Reply:
I believe it does maintain this constraint, just not the other direction - if the canonical hash is in the db then the header is. 
The header being in the db does not necessarily mean the hash is in the db (since there is a delay before `storeResult` is called.
This is what I meant by thinking it's ok, but that just means I haven't found an area that makes it not Ok yet

Reply:
Sounds good!